pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/GrpcBrokerRequestHandler.java

```diff
@@ -0,0 +1,162 @@
+    _streamingReduceService = new StreamingReduceService(config);
+  }
+
+  @Override
+  public void start() {
+  }
+
+  @Override
+  public synchronized void shutDown() {
+    _streamingReduceService.shutDown();
+  }
+
+  @Override
+  protected BrokerResponseNative processBrokerRequest(long requestId, BrokerRequest originalBrokerRequest,
+      @Nullable BrokerRequest offlineBrokerRequest, @Nullable Map<ServerInstance, List<String>> offlineRoutingTable,
+      @Nullable BrokerRequest realtimeBrokerRequest, @Nullable Map<ServerInstance, List<String>> realtimeRoutingTable,
+      long timeoutMs, ServerStats serverStats, RequestStatistics requestStatistics)
+      throws Exception {
+    assert offlineBrokerRequest != null || realtimeBrokerRequest != null;
+
+    String rawTableName = TableNameBuilder.extractRawTableName(originalBrokerRequest.getQuerySource().getTableName());
+    Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> responseMap = new HashMap<>();
+    // Making request to a streaming server response.
+
+    if (offlineBrokerRequest != null) {
+      // to offline servers.
+      assert offlineRoutingTable != null;
+      streamingQueryToPinotServer(requestId, _brokerId, rawTableName, TableType.OFFLINE, responseMap,
+          offlineBrokerRequest, offlineRoutingTable, timeoutMs, true, 1);
+    }
+    if (realtimeBrokerRequest != null) {
+      // to realtime servers.
+      assert realtimeRoutingTable != null;
+      streamingQueryToPinotServer(requestId, _brokerId, rawTableName, TableType.REALTIME, responseMap,
+          realtimeBrokerRequest, realtimeRoutingTable, timeoutMs, true, 1);
+    }
+    BrokerResponseNative brokerResponse = _streamingReduceService.reduceOnStreamResponse(
+        originalBrokerRequest, responseMap, timeoutMs, _brokerMetrics);
+    return brokerResponse;
+  }
+
+  /**
+   * Query pinot server for data table.
+   */
+  public void streamingQueryToPinotServer(final long requestId, final String brokerHost, final String rawTableName,
+      final TableType tableType, Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> responseMap,
+      BrokerRequest brokerRequest, Map<ServerInstance, List<String>> routingTable, long connectionTimeoutInMillis,
+      boolean ignoreEmptyResponses, int pinotRetryCount) {
+    // Retries will all hit the same server because the routing decision has already been made by the pinot broker
+    Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponseMap = new HashMap<>();
+    for (Map.Entry<ServerInstance, List<String>> routingEntry : routingTable.entrySet()) {
+      ServerInstance serverInstance = routingEntry.getKey();
+      List<String> segments = routingEntry.getValue();
+      String serverHost = serverInstance.getHostname();
+      int port = serverInstance.getGrpcPort();
+      // ensure concurrent request count from the same grpc server cannot exceed maximum
+      if (!_concurrentQueriesCountMap.containsKey(serverHost)) {
```

Comment:
Not thread safe.
Let's remove this map for now, and add it back if necessary when adding the throttling